File: scripts/TestRequest.kt
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 17+
//KOTLIN 1.7.20
//DEPS org.apache.commons:commons-text:1.9
//DEPS commons-codec:commons-codec:1.15
//DEPS org.jetbrains.kotlinx:kotlinx-coroutines-core:1.6.4
//DEPS io.ktor:ktor-client-core-jvm:2.1.2
//DEPS io.ktor:ktor-client-cio-jvm:2.1.2

import kotlinx.coroutines.*
import io.ktor.client.*
import io.ktor.client.engine.cio.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import org.apache.commons.codec.digest.HmacAlgorithms
import org.apache.commons.codec.digest.HmacUtils

fun main() = runBlocking {
    val signKey = "xxxxxxxxxxxxxxxxxxxxxxxxxx"
    val bodyText = """
        {
            "action": "push",
            "ref": "refs/heads/gh-pages"
        }
    """.trimIndent()

    val client = HttpClient(CIO)
    val response1 =
        client.post("http://localhost:8080/myproject") {
            headers {
                append(HttpHeaders.ContentType, "application/json")
                append(
                    "X-Hub-Signature-256",
                    "sha256=" + HmacUtils(HmacAlgorithms.HMAC_SHA_256, signKey).hmacHex(bodyText)
                )
            }
            setBody(bodyText)
        }
    println("HTTP ${response1.status}: ${response1.bodyAsText()}")
    //--------
    val response2 =
        client.post("http://localhost:8080/myproject") {
            headers {
                append(HttpHeaders.ContentType, "application/json")
                append(
                    "X-Hub-Signature",
                    "sha1=" + HmacUtils(HmacAlgorithms.HMAC_SHA_1, signKey).hmacHex(bodyText)
                )
            }
            setBody(bodyText)
        }
    println("HTTP ${response2.status}: ${response2.bodyAsText()}")
    //--------
    val response3 =
        client.post("http://localhost:8080/notAvailable") {
            headers {
                append(HttpHeaders.ContentType, "application/json")
                append(
                    "X-Hub-Signature",
                    "sha1=" + HmacUtils(HmacAlgorithms.HMAC_SHA_1, signKey).hmacHex(bodyText)
                )
            }
            setBody(bodyText)
        }
    println("HTTP ${response3.status}: ${response3.bodyAsText()}")
}

File: scripts/test-request.rs
#!/usr/bin/env -S cargo +nightly -Zscript --quiet
---
[package]
edition = "2024"

[dependencies]
reqwest = { version = "0.12", default-features = false, features = ["blocking"] }
hmac = "0.12"
sha1 = "0.10"
sha2 = "0.10"
hex = "0.4"
---

use hex::encode;
use hmac::{Hmac, Mac};
use reqwest::blocking::Client;
use sha1::Sha1;
use sha2::Sha256;
use std::error::Error;
use std::io::Error as IoError;

type HmacSha256 = Hmac<Sha256>;
type HmacSha1 = Hmac<Sha1>;

fn generate_hmac_sha256(body: &str, secret: &str) -> Result<String, Box<dyn Error>> {
    let mut mac = HmacSha256::new_from_slice(secret.as_bytes())
        .map_err(|error| IoError::other(format!("invalid HMAC key: {error}")))?;
    mac.update(body.as_bytes());
    let result = mac.finalize();
    Ok(format!("sha256={}", encode(result.into_bytes())))
}

fn generate_hmac_sha1(body: &str, secret: &str) -> Result<String, Box<dyn Error>> {
    let mut mac = HmacSha1::new_from_slice(secret.as_bytes())
        .map_err(|error| IoError::other(format!("invalid HMAC key: {error}")))?;
    mac.update(body.as_bytes());
    let result = mac.finalize();
    Ok(format!("sha1={}", encode(result.into_bytes())))
}

fn main() -> Result<(), Box<dyn Error>> {
    let sign_key = "xxxxxxxxxxxxxxxxxxxxxxxxxx";
    let body_text = r#"{
    "action": "push",
    "ref": "refs/heads/gh-pages"
}"#;

    let client = Client::new();

    println!("Test 1: GET /");
    let response = client.get("http://localhost:8080/").send()?;
    println!("HTTP {}: {}", response.status(), response.text()?);
    println!();

    println!("Test 2: POST /myproject with SHA256");
    let signature_sha256 = generate_hmac_sha256(body_text, sign_key)?;
    let response = client
        .post("http://localhost:8080/myproject")
        .header("Content-Type", "application/json")
        .header("X-Hub-Signature-256", signature_sha256)
        .body(body_text)
        .send()?;
    println!("HTTP {}: {}", response.status(), response.text()?);
    println!();

    println!("Test 3: POST /myproject with SHA1");
    let signature_sha1 = generate_hmac_sha1(body_text, sign_key)?;
    let response = client
        .post("http://localhost:8080/myproject")
        .header("Content-Type", "application/json")
        .header("X-Hub-Signature", signature_sha1)
        .body(body_text)
        .send()?;
    println!("HTTP {}: {}", response.status(), response.text()?);
    println!();

    println!("Test 4: POST /notAvailable");
    let signature = generate_hmac_sha1(body_text, sign_key)?;
    let response = client
        .post("http://localhost:8080/notAvailable")
        .header("Content-Type", "application/json")
        .header("X-Hub-Signature", signature)
        .body(body_text)
        .send()?;
    println!("HTTP {}: {}", response.status(), response.text()?);

    Ok(())
}
